src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java

```diff
@@ -0,0 +1,237 @@
+package emu.grasscutter.game.managers.blossom;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import emu.grasscutter.Grasscutter;
+import emu.grasscutter.data.GameData;
+import emu.grasscutter.data.GameDepot;
+import emu.grasscutter.data.common.ItemParamData;
+import emu.grasscutter.data.excels.RewardPreviewData;
+import emu.grasscutter.game.entity.EntityGadget;
+import emu.grasscutter.game.entity.GameEntity;
+import emu.grasscutter.game.entity.gadget.GadgetWorktop;
+import emu.grasscutter.game.inventory.GameItem;
+import emu.grasscutter.game.player.Player;
+import emu.grasscutter.game.world.Scene;
+import emu.grasscutter.game.world.SpawnDataEntry;
+import emu.grasscutter.game.world.SpawnDataEntry.SpawnGroupEntry;
+import emu.grasscutter.net.proto.BlossomBriefInfoOuterClass;
+import emu.grasscutter.net.proto.VisionTypeOuterClass;
+import emu.grasscutter.server.packet.send.PacketBlossomBriefInfoNotify;
+import emu.grasscutter.utils.Utils;
+import it.unimi.dsi.fastutil.ints.Int2ObjectLinkedOpenHashMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+
+public class BlossomManager {
+    public BlossomManager(Scene scene) {
+        this.scene = scene;
+    }
+
+    private final Scene scene;
+    private final List<BlossomActivity> blossomActivities = new ArrayList<>();
+    private final List<BlossomActivity> activeChests = new ArrayList<>();
+    private final List<EntityGadget> createdEntity = new ArrayList<>();
+
+    private final List<SpawnDataEntry> blossomConsumed = new ArrayList<>();
+
+    public void onTick(){
+        synchronized (blossomActivities){
+            var it = blossomActivities.iterator();
+            while(it.hasNext()){
+                var active = it.next();
+                active.onTick();
+                if (active.getPass()) {
+                    EntityGadget chest = active.getChest();
+                    scene.addEntity(chest);
+                    scene.setChallenge(null);
+                    activeChests.add(active);
+                    it.remove();
+                }
+            }
+        }
+    }
+
+    public void recycleLeyLineGadgetEntity(List<GameEntity> entities){
+        for(var entity : entities){
+            if(entity instanceof EntityGadget gadget){
+                createdEntity.remove(gadget);
+            }
+        }
+        notifyIcon();
+    }
+
+    public void initBlossom(EntityGadget gadget){
+        if(createdEntity.contains(gadget)){
+            return;
+        }
+        if(blossomConsumed.contains(gadget.getSpawnEntry())){
+            return;
+        }
+        var id = gadget.getGadgetId();
+        if(BlossomType.valueOf(id)==null){
+            return;
+        }
+        gadget.buildContent();
+        gadget.setState(204);
+        int worldLevel = getWorldLevel();
```

Comment:
we should check open state for the player(level>2),and the level should be read from file `BlossomOpenExcelConfigData.json`

Reply:
Actually the unlock conditions are under `refreshCondVec` in `Resources/ExcelBinOutput/BlossomRefreshExcelConfigData.json` - it's AR8 for the Mondstadt EXP leyline, AR12 for Mondstadt mora leyline, AR18 for both Liyue leylines, AR30 for both Inazuma leylines.

Copying the official spawns will be a bit more work after this PR.